images/virtualization-artifact/pkg/controller/vmclass/internal/watcher/vm_watcher.go

```
@@ -0,0 +1,79 @@
+/*
+Copyright 2025 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watcher
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+type VirtualMachinesWatcher struct{}
+
+func NewVirtualMachinesWatcher() *VirtualMachinesWatcher {
+	return &VirtualMachinesWatcher{}
+}
+
+func (w *VirtualMachinesWatcher) Watch(mgr manager.Manager, ctr controller.Controller) error {
+	return ctr.Watch(
+		source.Kind(mgr.GetCache(), &virtv2.VirtualMachine{}),
+		handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []reconcile.Request {
+			vm, ok := obj.(*virtv2.VirtualMachine)
+			if !ok {
+				return nil
+			}
+
+			c := mgr.GetClient()
+			vmc := &virtv2.VirtualMachineClass{}
+			err := c.Get(ctx, types.NamespacedName{
+				Name: vm.Spec.VirtualMachineClassName,
+			}, vmc)
+			if err != nil {
+				return nil
```

Dont ignore errors, never
Use FetchObject